src/inference/src/dev/plugin_config.cpp

```diff
@@ -0,0 +1,319 @@
+    if (m_user_properties.find(name) != m_user_properties.end()) {
+        return m_user_properties.at(name);
+    }
+
+    auto option = get_option_ptr(name);
+    OPENVINO_ASSERT((allowed_visibility & option->get_visibility()) == option->get_visibility(), "Couldn't get unknown property: ", name);
+
+    return option->get_any();
+}
+
+void PluginConfig::set_property(const ov::AnyMap& config) {
+    OPENVINO_ASSERT(!m_is_finalized, "Setting property after config finalization is prohibited");
+
+    for (auto& [name, val] : config) {
+        get_option_ptr(name)->set_any(val);
+    }
+}
+
+void PluginConfig::set_user_property(const ov::AnyMap& config, OptionVisibility allowed_visibility, bool throw_on_error) {
+    OPENVINO_ASSERT(!m_is_finalized, "Setting property after config finalization is prohibited");
+
+    for (auto& [name, val] : config) {
+        auto option = get_option_ptr(name);
+        if ((allowed_visibility & option->get_visibility()) != option->get_visibility()) {
+            if (throw_on_error)
+                OPENVINO_THROW("Couldn't set unknown property: ", name);
+            else
+                continue;
+        }
+        if (!option->is_valid_value(val)) {
+            if (throw_on_error)
+                OPENVINO_THROW("Invalid value: ", val.as<std::string>(), " for property: ",  name, "\nProperty description: ", get_help_message(name));
+            else
+                continue;
+        }
+
+        m_user_properties[name] = val;
+    }
+}
+
+void PluginConfig::finalize(const IRemoteContext* context, const ov::Model* model) {
+    if (m_is_finalized)
+        return;
+
+    if (model)
+        apply_model_specific_options(context, *model);
+
+    // Copy internal properties before applying hints to ensure that
+    // a property set by hint won't be overriden by a value in user config.
+    // E.g num_streams=AUTO && hint=THROUGHPUT
+    // If we apply hints first and then copy all values from user config to internal one,
+    // then we'll get num_streams=AUTO in final config while some integer number is expected.
+    for (const auto& prop : m_user_properties) {
+        auto& option = m_options_map.at(prop.first);
+        option->set_any(prop.second);
+    }
+
+    finalize_impl(context);
+
+#ifdef ENABLE_DEBUG_CAPS
+    apply_env_options();
+#endif
+
+    // Clear properties after finalize_impl to be able to check if a property was set by user during plugin-side finalization
+    m_user_properties.clear();
+
+    m_is_finalized = true;
+}
+
+bool PluginConfig::visit_attributes(ov::AttributeVisitor& visitor) {
+    visitor.on_attribute("m_user_properties", m_user_properties);
+    for (auto& prop : m_options_map) {
+        visitor.on_attribute(prop.first + "__internal", prop.second);
+    }
+
+    return true;
+}
+
+void PluginConfig::apply_env_options() {
+    ov::AnyMap env_properties = read_env();
+    cleanup_unsupported(env_properties);
+    for (auto& [name, val] : env_properties) {
```

Comment:
It looks like debug message should be removed.
Add helper to or macro to print options in debug mode?

Reply:
These messages are left intentionally. They indicate that some values are really applied from env/config. Those are visible for the build with debug caps only, so it's not an issue.